### ISSUE TITLE
Sections support in Blazor

### ISSUE THREAD

**Issue by akorchev** [dotnet/aspnetcore#28182]
Basically the same as https://github.com/dotnet/aspnetcore/issues/10131 which has been closed in in favour of #10452 which was in turn closed in favour of #10450 (influencing the `<head>`). 

Having the ability to embed content from the page in the layout in more than one placeholder is a common need. A typical use case is to display the page name in the "header" of the application which is usually defined in the layout. Another use case is to display personalised content for the currently logged user.

A lot of other technologies support this feature and I think it would be a great addition to Blazor. A few examples:

- ASP.NET WebForms has [ContentPlaceholder](https://docs.microsoft.com/en-us/aspnet/web-forms/overview/older-versions-getting-started/master-pages/multiple-contentplaceholders-and-default-content-cs)
- ASP.NET MVC has [sections](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/layout?view=aspnetcore-5.0)
- Web Pages also have sections
- Angular has [router outlets](https://angular.io/api/router/RouterOutlet#description)
- React Router also [supports multiple components](https://gist.github.com/kitdemonio/33be59d4ab597f2679e83d68e1b63842)

Thank you for your consideration!

**Comment by SteveSandersonMS** [comment]
One of the reasons we haven't got this as a baked-in feature yet is that it's possible to implement something like this in your own application code. If there's enough consensus that some single pattern takes care of almost all requirements, we might well consider putting it into the framework. But until then, it would help if people were able to try out ways of doing this and report back on what's working well or not well for them.

To get started, here's one way to do it:

 * Add the three classes (`SectionContent`, `SectionOutlet`, and `SectionRegistry`) from this Gist to your project: https://gist.github.com/SteveSandersonMS/4f08efe2ad32178add12bfa3eb6e4559
 * In your layout, or in fact in any component (but normally your layout), use `<SectionOutlet>` with a name to define where you want some content to go. For example, you could add the following into the default template's `MainLayout.razor`:

```razor
<div class="top-row px-4">
    <SectionOutlet Name="topbar" />
    <a href="https://docs.microsoft.com/aspnet/" target="_blank">About</a>
</div>
```

 * In any other page, or in fact in any component (but normally in a page), use `<SectionContent>` with a name to define some content to be rendered. For example, you change `Counter.razor` to render the following:

```razor
<h1>Counter</h1>

<p>Current count: @currentCount</p>

<SectionContent Name="topbar">
    <button class="btn btn-primary" @onclick="IncrementCount">Increment counter</button>
</SectionContent>
```

Now when the user navigates to "Counter", they will have a button in the navigation bar that updates the count. And when they navigate to any other page, that button will disappear.

![image](https://user-images.githubusercontent.com/1101362/100469637-85232000-30ce-11eb-926e-23a070ee7c25.png)

Of course, each page could provide its own different content for `topbar`, and you can have any number of different named sections.

Technically you can even use this approach to supply content to the `<head>` element in your page (e.g., to have a page-specific `<title>`) - I checked this works. It's a bit awkward because you do need to migrate everything inside `<html>` to be part of your `App.razor` root component so the `<head>` is rendered by Blazor. But it can be done. If this sort of approach pleases people we can look into ways of making it more natural in .NET 6.

**Comment by akorchev** [comment]
Thank you @SteveSandersonMS ! This is a viable solution.

> One of the reasons we haven't got this as a baked-in feature yet is that it's possible to implement something like this in your own application code

Indeed it is. But the implementation is so simple and small that you can consider including it in Blazor (or another official Nuget package that people can use). Thanks again for the code and general idea. I will use it in my projects.

**Comment by akorchev** [comment]
To anybody trying @SteveSandersonMS's solution don't make the same mistake as me. Do not forget to include `@Body` in your layout! I assumed it was not needed when using SectionOutlet and was wondering why nothing worked. Rendering the Body is still needed in order to actually run the code in the SectionContent and register it with the SectionRegistry.

**Comment by haefele** [comment]
Is this still being considered for .NET 7 ?
As far as I can tell the outlet-code is already included in Blazor - just internal.

**Comment by SteveSandersonMS** [comment]
It's in `.NET 7 Planning` so it may or may not make it in depending on the overall workload and how it's prioritized relative to other things. You're correct that the implementation is basically already done - the remaining cost is around ensuring the API is correct for long-term support, figuring out if there might be any problems caused by incorrect usage, docs, perhaps more tests for other use cases, etc.

**Comment by SocVi100** [comment]
Any news about this?
I've used the above code to create Sections but I'd like to use the official bits...

**Comment by sajjadarashhh** [comment]
@javiercn thanks to mention #42880 here.
we can hope to have solution for this problem in blazor and .net 7!?

**Comment by javiercn** [comment]
@sajjadarashhh not for .NET 7 for sure.

**Comment by wangkanai** [comment]
@danroth27 would this issue be implemented as public in the .NET7 planning or its going to be push back to .NET8?

Thanks and best regards

**Comment by javiercn** [comment]
@wangkanai this is not something we are doing for .NET 7.0. 

We will consider it for 8.0, however, we want to better understand concrete scenarios people have around the feature before we decide to include it in the framework. Specifically with comparisons between the code you have to write today and the code you would write with the feature in place.

**Comment by wangkanai** [comment]
Okay, I am writing a demo project blazor that would need this feature. https://github.com/wangkanai/wangkanai/tree/main/Tabler
I’m making the blazor version of the Tabler CCS framework.
https://tabler.wangkanai.com/

What I’m looking to do the change body css class, similar to what you can do with `<PageTitle>`

**Comment by javiercn** [comment]
@wangkanai I am not sure we have enough detail to understand what you are trying to accomplish. Could you provide a concrete code snippet?

**Comment by wangkanai** [comment]
Okay, Let me design something up quickly.

**Comment by wangkanai** [comment]
I would like at extend the Sections to public and I would use it update `<Body>` class with example code below. By design it would work just the like the blazor `<PageTitle>` component.

![image](https://user-images.githubusercontent.com/10666633/189178955-47a9667e-bcb2-47b4-81ce-9494934227c5.png)
Reference: [Box layout](https://github.com/wangkanai/wangkanai/blob/c3ec378a71c87836fd1dd7a15be9e1660b756b0d/Tabler/demo/Client/Pages/Layouts/Boxed.razor)

**Comment by SteveSandersonMS** [comment]
@wangkanai Thanks for providing this code sketch. This helps to clarify, because what you're looking for doesn't really resemble the sections proposal we have. You need some way to add CSS classes into an existing external element, which is not something addressed by sections.

Instead I recommend you simply use JS interop to toggle the class on the `<body>` element.

**Comment by SocVi100** [comment]
I'll expose my use case, just in case it serves as an example:
I have a main layout with a header. This header has a button on the left to open the main menu (a sidenav panel on the main layout also).
But this header also contains the title of the current page and one or more buttons on the left that depend on the contents of the page being opened.
So, I created a section inside this header using the above code, and I fill it's different content on every page of the app. Otherwise I'd have to replicate the header and the main layout structure on every page...

**Comment by yugabe** [comment]
Just wanted to add my few cents. I'm making a CMS-like engine, where different pages can render to different parts of a page. Namely, on the top, there are breadcrumbs, and on the right, there is a table of contents. The body of the page is the content itself. The layout can define the sections.

I'm not a big fan of magic strings, I'd rather pass discrete references to objects, so I've made the following, using the example of putting a table of contents on the page in a different place. It's similar (closest) in concept to MVC's .cshtml sections.


<code>Section.cs</code>
```cs
public sealed class Section<TOutlet> : IComponent where TOutlet : SectionOutlet
{
    [Parameter, EditorRequired]
    public RenderFragment ChildContent { get; set; } = null!;

    [Parameter, EditorRequired]
    public TOutlet Outlet { get; set; } = null!;

    void IComponent.Attach(RenderHandle renderHandle)
    {
        // The section isn't rendered where it is defined, so the RenderHandle is discarded here by design.
    }

    public Task SetParametersAsync(ParameterView parameters)
    {
        ChildContent = parameters.GetValueOrDefault<RenderFragment>(nameof(ChildContent))!;
        Outlet = parameters.GetValueOrDefault<TOutlet>(nameof(Outlet))!;
        Outlet.Render(ChildContent);
        return Task.CompletedTask;
    }
}
```

<code>SectionOutlet.cs</code>
```cs
public class SectionOutlet : IComponent
{
    public RenderHandle Handle { get; private set; }
    public void Attach(RenderHandle renderHandle) => Handle = renderHandle;
    public virtual Task SetParametersAsync(ParameterView parameters) => Task.CompletedTask;
    public void Render(RenderFragment renderFragment) => Handle.Render(renderFragment);
}
```

Usage is as expected: you need to create an outlet somewhere (higher up the hierarchy), hold its reference, then define a section in the body of your page referring to the reference held, something like:

<code>App.razor</code>
```razor

<CascadingValue Value="this" IsFixed="true">
    <!-- Router, layout, etc. -->

    <div class="table-of-contents">
        <SectionOutlet @ref="TableOfContents">
    </div>
</CascadingValue>

@code {
    public SectionOutlet? TableOfContents { get; private set; }
}
```

<code>MyPage.razor</code>
```razor
@page /mypage

<Section Outlet="App.TableOfContents">
    <h1>Lorem ipsum dolor</h1>
    <h2>Lorem ipsum dolor</h3>
    <h3>Lorem ipsum dolor</h3>
</Section>

@code {
    [CascadingParameter]
    public App App { get; set; }
}
```

This seems to work for my scenario very well. I wonder if I'm missing something or if this will cause issues in the future, but I would like to see this implemented in the framework as well. One obvious drawback of this direct implementation is if there are multiple Section instances using the same Outlet, they'll keep overwriting each others' contents. I'm not sure how this could be done otherwise (except for throwing in such cases when multiple section try to write to the same outlet).

The other one using the breadcrumbs I'm currently in the middle of developing, as it shouldn't use arbitrary content, but rather should use a model provided by the page itself; so it's possibly better to implement it in some other way, or it requires maybe one or two more levels of abstraction.

**Comment by mkalinski93** [comment]
It also would be nice to check if the section is available or has content similar to [IsSectionDefined](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.razor.razorpage.issectiondefined?view=aspnetcore-7.0)

**Comment by Flachdachs** [comment]
@yugabe 
I like your version, it's very short and clean. I don't know if you have improved it in the meantime, but I found two issues.

1. When a component is disposed that uses a `Section` the `SectionOutlet` keeps its content. To fix this issue you can implement `IDisposable` in Section and call `Outlet?.Render(_=>{});` in the `Dispose` method.
2. When a Layout page has outlets above and below `@Body` a NullReferenceException is thrown for the outlets below `@Body` because they don't exists yet when the Section's `SetParametersAsync` of a component in the `@Body` is called. The `?` in `Outlet?.Render(ChildContent);` fixes the exception, but the `ChildContent` isn't rendered until it get's a rerendering.

**Comment by yugabe** [comment]
@Flachdachs 
Thanks a lot for the comment! My original solution was just a proof of concept I was hoping the team could use when they implemented the in-box solution. However, it seems they opted to make public the previously private version, so this wasn't needed. I'm not sure I'll move unconditionally to that solution because I think this one has a few advantages over that.

Both of your problems I also encountered myself and solved them, but in the meantime the solution got more verbose and complicated (I actually created a third component called `RenderSection`, which handles rendering the `SectionOutlet`), so I didn't keep the above code up-to-date. I'll make an update to the code above though, so anyone stumbling on it can use it as is. Thank you!
